File: client_test.js
var createNode = require('secret-handshake/net')
var fs = require('fs')
var pull = require('pull-stream')


function readKeyF (fname) {
  var tmpobj = JSON.parse(fs.readFileSync(fname).toString())
  return {
    'publicKey': new Buffer(tmpobj.publicKey, 'base64'),
    'secretKey': new Buffer(tmpobj.secretKey, 'base64'),
  }
}

var alice = readKeyF('key.alice.json')
var bob = readKeyF('key.bob.json')

var alice = createNode({
  keys: alice,
  appKey: new Buffer('IhrX11txvFiVzm+NurzHLCqUUe3xZXkPfODnp7WlMpk=', 'base64'),
})
console.log('dialing...')
var stream = alice.connect({port: 8978, host: 'localhost', key: bob.publicKey}, function (err,stumb) {
  if (err) throw err
  console.log('dialed!')
  pull(pull.values(['hello', 'world']), stumb)
})

File: client_test.go
package shs

import (
	"encoding/base64"
	"encoding/json"
	"os"
	"testing"

	"github.com/cryptix/go/logging/logtest"
	"github.com/cryptix/go/proc"
)

func check(err error) {
	if err != nil {
		panic(err)
	}
}

func mustLoadKeyPair(fname string) EdKeyPair {
	f, err := os.Open(fname)
	check(err)

	var t struct {
		PublicKey, SecretKey string
	}
	check(json.NewDecoder(f).Decode(&t))

	pubClient, err := base64.StdEncoding.DecodeString(t.PublicKey)
	check(err)

	secSrv, err := base64.StdEncoding.DecodeString(t.SecretKey)
	check(err)

	var kp EdKeyPair
	copy(kp.Public[:], pubClient)
	copy(kp.Secret[:], secSrv)
	return kp
}

func TestClient(t *testing.T) {
	w := logtest.Logger("server_test.js", t)
	server, err := proc.StartStdioProcess("node", w, "server_test.js")
	if err != nil {
		t.Fatal(err)
	}

	appKey, err := base64.StdEncoding.DecodeString("IhrX11txvFiVzm+NurzHLCqUUe3xZXkPfODnp7WlMpk=")
	if err != nil {
		t.Fatal(err)
	}

	kpBob := mustLoadKeyPair("key.bob.json")
	kpAlice := mustLoadKeyPair("key.alice.json")

	clientState, err := NewClientState(appKey, kpAlice, kpBob.Public)
	if err != nil {
		t.Fatal("error making server state:", err)
	}

	if err := Client(*clientState, server); err != nil {
		t.Fatal(err)
	}

	if err := server.Close(); err != nil {
		t.Fatal(err)
	}
}
